### PR TITLE
[4846][ADD] purchase_cancel_product_archive

### PR DESCRIPTION
**Pull request by AungKoKoLin1997**
[4846](https://www.quartile.co/web#id=4846&cids=3&model=project.task&view_type=form)

### CODE REVIEW

**Review by yostashiro — COMMENTED** [review]
Would it make more sense to extend the write() method of purchase.order to do the archive? We can also revive the product when the state is reset to draft. I prefer not to add a cron when it's not necessary. @kanda999 What do you think?

We can use the current code to clean up the existing records in post_init_hook.

**Comment by yostashiro** [comment]
> Would it make more sense to extend the write() method of purchase.order to do the archive? We can also revive the product when the state is reset to draft. I prefer not to add a cron when it's not necessary. @kanda999 What do you think?

In case we go for this, I suggest the module name to be purchase_cancel_product_archive.

**Comment by kanda999** [comment]
@AungKoKoLin1997 
Would you consider yoshi-san's suggestion?　There is no need to do this process in cron.
Archives the product when the purchase.order is canceled, and unarchives it when it is reversed to a quote.

**Comment by AungKoKoLin1997** [comment]
> @AungKoKoLin1997 Would you consider yoshi-san's suggestion?　There is no need to do this process in cron. Archives the product when the purchase.order is canceled, and unarchives it when it is reversed to a quote.

I agree. Lets go with yoshi's suggestion.

File: purchase_cancel_product_archive/models/purchase_order.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        if not vals.get("state"):
+            return super().write(vals)
+        for rec in self:
+            products = rec.order_line.mapped("product_id")
+            if rec.state == "cancel":
+                products.write({"active": True})
+                products.mapped("product_tmpl_id").write({"active": True})
+            elif vals["state"] == "cancel":
+                products.write({"active": False})
+                products.mapped("product_tmpl_id").write({"active": False})
```
Comment (yostashiro) [inline purchase_cancel_product_archive/models/purchase_order.py]: The logic looks ~incorrect~/suboptimal.  I suggest you extend `button_cancel()` and `button_draft()` instead.

**Review by yostashiro — CHANGES_REQUESTED** [review]

File: purchase_cancel_product_archive/hooks.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# Copyright 2024 Quartile (https://www.quartile.co)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html)
+
+from odoo import SUPERUSER_ID, api
+
+
+def post_init_hook(cr, registry):
+    env = api.Environment(cr, SUPERUSER_ID, {})
+    cancel_pos = env["purchase.order"].search([("state", "=", "cancel")])
+    products = cancel_pos.mapped("order_line.product_id")
```
Comment (yostashiro) [inline purchase_cancel_product_archive/hooks.py:10]: Just to add some precaution; there could be some rare situations where a PO is cancelled but the same product is used in another PO. Haven't checked if such case actually exists.
```suggestion
    products = cancel_pos.mapped("order_line.product_id")
    products = products.filtered(lambda x: x.quantity_available <= 0.0)
```

**Review by yostashiro — APPROVED** [review]

**Review by kanda999 — APPROVED** [review]

**Pull request merged**